[WIP] Docs update for Alpha JobBackoffLimitPerIndex [placeholder PR] (#41921)
kubernetes/website: content/en/docs/concepts/workloads/controllers/job.md

By default, each pod failure is counted towards the `.spec.backoffLimit` limit,
see [pod backoff failure policy](#pod-backoff-failure-policy). However, you can
customize handling of pod failures by setting the Job's [pod failure policy](#pod-failure-policy).
Additionally, you can choose to count the pod failures independently for each

> index of an Indexed Job by setting the `.spec.backoffLimitPerIndex` field

Reviewer: ```suggestion
index of an [indexed](#completion-mode) Job by setting the `.spec.backoffLimitPerIndex` field
```

from failed Jobs is not lost inadvertently.
{{< /note >}}

### Backoff limit per index {#backoff-limit-per-index}

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

{{< note >}}
You can only configure the backoff limit per index for a Job, if you
have the `JobBackoffLimitPerIndex` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)

> enabled in your cluster. This feature is only available for Indexed Jobs, which

Reviewer: ```suggestion
enabled in your cluster. A per-index backoff limit is also only available for indexed Jobs that
```

use the `restartPolicy: Never`.
{{< /note >}}

> When you run an Indexed Job, you may want to handle pod failures for different

Reviewer: ```suggestion
When you run an [indexed](#completion-mode) Job, you may want to handle pod failures for different
```

indexes as counted independently. To do so, set the `.spec.backoffLimitPerIndex`

> to specify the the maximal number of pod failures per index. When the number is

Reviewer: ```suggestion
to specify the the maximal number of pod failures per index. For example, to allow exactly
one retry for each indexed subtask in a Job and to fail early if the retry also fails, set `.spec.backoffLimitPerIndex` to 1 and set `.spec.maxFailedIndexes` to 0.

When the per-index backoff limit is
```

exceeded for an index, the index is marked as failed in the
`.status.failedIndexes` field. Note that, a failing index does not interrupt
execution of other indexes. You can additionally limit the maximal number of
indexes marked failed by setting the `.spec.maxFailedIndexes` field. In that case,
the the number of failed indexes is exceeded, then the entire Job is marked

> failed and execution all pods and indexed is terminated.

Reviewer: ```suggestion
failed; marking that failure triggers termination of all running Pods for that Job.
```

> When all indexes finishes for a Job using this feature, and there is at least
> one failed index, then the entire Job is marked as failed, even if some indexes
> are finished successfully.

Reviewer: ```suggestion
If all indexes finish for a Job where you specified a per-index backoff limit, and at
least one of those indexes did fail, but there wasn't an early termination (due to
reaching `.spec.maxFailedIndexes`), then at this point the Kubernetes control plane
marks the overall Job as failed. The overall Job gets marked as failed even if some,
perhaps nearly all, of the indexes were processed successfully.
```

> Here is a manifest for an Indexed Job that defines a `backoffLimitPerIndex`:

Reviewer: ```suggestion
Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:
```

from failed Jobs is not lost inadvertently.
{{< /note >}}

### Backoff limit per index {#backoff-limit-per-index}

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

{{< note >}}
You can only configure the backoff limit per index for a Job, if you
have the `JobBackoffLimitPerIndex` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
enabled in your cluster. A per-index backoff limit is also only available for indexed Jobs that
use the `restartPolicy: Never`.
{{< /note >}}

When you run an [indexed](#completion-mode) Job, you may want to handle pod failures for different
indexes as counted independently. To do so, set the `.spec.backoffLimitPerIndex`
to specify the the maximal number of pod failures per index. For example, to allow exactly

> one retry for each indexed subtask in a Job and to fail early if the retry also fails, set `.spec.backoffLimitPerIndex` to 1 and set `.spec.maxFailedIndexes` to 0.

Reviewer 2: This is too specific at this point. Maybe you can have a paragraph at the end of the section saying:

```
Here are some sample configurations you can set for your indexed Jobs:
- `backoffLimitPerIndex=1` and `maxFailedIndexes=0`, meaning that...
```

Author: I think this is something we could consider moving to the tasks page. I'm thinking about "Execute all indexes" title which would correspond with the original Issue: https://github.com/kubernetes/kubernetes/issues/109712.

Author: Resolving as the original comment is addressed, and the need to have a new tasks page is discussed in a separate thread: https://github.com/kubernetes/website/pull/41921#discussion_r1272202923

When the per-index backoff limit is
exceeded for an index, the index is marked as failed in the
`.status.failedIndexes` field. Note that, a failing index does not interrupt
execution of other indexes. You can additionally limit the maximal number of
indexes marked failed by setting the `.spec.maxFailedIndexes` field. In that case,
the the number of failed indexes is exceeded, then the entire Job is marked
failed; marking that failure triggers termination of all running Pods for that Job.

> If all indexes finish for a Job where you specified a per-index backoff limit, and at

Reviewer 2: ```suggestion
Once all indexes finish for a Job where you specified a per-index backoff limit, if at
```

least one of those indexes did fail, but there wasn't an early termination (due to

> reaching `.spec.maxFailedIndexes`), then at this point the Kubernetes control plane

Reviewer 2: it doesn't matter if it reaches the maxFailedIndexes or not. All that matters is that at least one Index failed.

Author: Yes, remove that part.

> reaching `.spec.maxFailedIndexes`), then at this point the Kubernetes control plane

Reviewer 2: ```suggestion
reaching `.spec.maxFailedIndexes`), the Kubernetes control plane
```

marks the overall Job as failed. The overall Job gets marked as failed even if some,

> perhaps nearly all, of the indexes were processed successfully.

Reviewer 2: worth mentioning the `completedIndexes` field.

Author: Mentioned above now, as it seems to fit even better there.

Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:

{{< codenew file="/controllers/job-backoff-limit-per-index-example.yaml" >}}

In the example above, the Job controller will allow for one restart for each
of the indexes. When the total number of failed indexes exceeds 5, then

> the entire Job is terminated.

Reviewer 2: ```
Once the job is finished, the the Job status will look as follows:

kubectl get -o yaml job job-name

status:
  failedIndexes: ".."
  completedIndexes: ".."
  conditions:
  - ...
```

Author: Done

executed pods, are recorded in the `.status.completedIndexes` field, as for
regular Indexed jobs.

Note that, a failing index does not interrupt execution of other indexes.
You can additionally limit the maximal number of indexes marked failed by
setting the `.spec.maxFailedIndexes` field. In that case,
the the number of failed indexes is exceeded, then the entire Job is marked
failed; marking that failure triggers termination of all running Pods for that Job.

Once all indexes finish for a Job where you specified a per-index backoff limit,
if at least one of those indexes did fail, the Kubernetes control plane
marks the overall Job as failed. The Job gets marked as failed even if some,
perhaps nearly all, of the indexes were processed successfully.

Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:

{{< codenew file="/controllers/job-backoff-limit-per-index-example.yaml" >}}

In the example above, the Job controller will allow for one restart for each
of the indexes. When the total number of failed indexes exceeds 5, then
the entire Job is terminated.

Once the job is finished, the the Job status will look as follows:

```sh

> kubectl get -o yaml job job-name

Reviewer: The Job is called `job-backoff-limit-per-index-example`, right?

Author: correct, fixed

```

```yaml
  status:
    completedIndexes: 1,3,5,7,9
    conditions:
    - message: Job has failed indexes
      reason: FailedIndexes
      status: "True"
      type: Failed

>     failed: 10

Author: Actually, there reason is 2 failed pods per 5 failed indexes. I have added now a comment to explain.

>     failedIndexes: 0,2,4,6,8

Reviewer: could move the `failedIndexes` to come after `completedIndexes`. Slightly different from the actual output but more helpful I think, and anyway this isn't the whole object.

enabled in your cluster. A per-index backoff limit is only available for indexed Jobs that
use the `restartPolicy: Never`.
{{< /note >}}

When you run an [indexed](#completion-mode) Job, you can choose to handle retries
for pod failures independently for each index. To do so, set the
`.spec.backoffLimitPerIndex` to specify the the maximal number of pod failures
per index.

When the per-index backoff limit is exceeded for an index, the index is marked as failed in the
`.status.failedIndexes` field. The succeeded indexes, those with a successfully
executed pods, are recorded in the `.status.completedIndexes` field, as for
regular Indexed jobs.

Note that, a failing index does not interrupt execution of other indexes.
You can additionally limit the maximal number of indexes marked failed by
setting the `.spec.maxFailedIndexes` field. In that case,
the the number of failed indexes is exceeded, then the entire Job is marked
failed; marking that failure triggers termination of all running Pods for that Job.

Once all indexes finish for a Job where you specified a per-index backoff limit,
if at least one of those indexes did fail, the Kubernetes control plane

> marks the overall Job as failed. The Job gets marked as failed even if some,

Reviewer 2: maybe more specific: sets a Failed condition in the status

Author: Makes sense, clarified here and in the previous paragraph.

perhaps nearly all, of the indexes were processed successfully.

Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:

{{< codenew file="/controllers/job-backoff-limit-per-index-example.yaml" >}}

In the example above, the Job controller will allow for one restart for each
of the indexes. When the total number of failed indexes exceeds 5, then
the entire Job is terminated.

Once the job is finished, the the Job status will look as follows:

```sh
kubectl get -o yaml job job-backoff-limit-per-index-example
```

```yaml
  status:
    completedIndexes: 1,3,5,7,9
    failedIndexes: 0,2,4,6,8
    succeeded: 5          # 1 succeeded pod for each of 5 succeeded indexes
    failed: 10            # 2 failed pods (1 retry) for each of 5 failed indexes
    conditions:
    - message: Job has failed indexes
      reason: FailedIndexes
      status: "True"
      type: Failed
```

Additionally, you may want to use the per-index backoff along with the
custom [pod failure policy](#pod-failure-policy). In particular, when using

> per-index backoff there is a new `FailIndex` action available which allows to

Reviewer 2: ```suggestion
per-index backoff there is a new `FailIndex` action available which allows you to
```

setting the `.spec.maxFailedIndexes` field.
When the number of failed indexes exceeds the `maxFailedIndexes` field, the
Job controller triggers termination of all remaining running Pods for that Job.
Once all pods are terminated, the entire Job is marked failed by the Job
controller, by setting the Failed condition in the Job status.

Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:

{{< codenew file="/controllers/job-backoff-limit-per-index-example.yaml" >}}

In the example above, the Job controller allows for one restart for each
of the indexes. When the total number of failed indexes exceeds 5, then
the entire Job is terminated.

Once the job is finished, the the Job status looks as follows:

```sh
kubectl get -o yaml job job-backoff-limit-per-index-example
```

```yaml
  status:
    completedIndexes: 1,3,5,7,9
    failedIndexes: 0,2,4,6,8
    succeeded: 5          # 1 succeeded pod for each of 5 succeeded indexes
    failed: 10            # 2 failed pods (1 retry) for each of 5 failed indexes
    conditions:
    - message: Job has failed indexes
      reason: FailedIndexes
      status: "True"
      type: Failed
```

> Additionally, you may want to use the per-index backoff along with the
> custom [pod failure policy](#pod-failure-policy). When using

Reviewer 2: ```suggestion
Additionally, you may want to use the per-index backoff along with a
[pod failure policy](#pod-failure-policy). When using
```

Author: Done

from failed Jobs is not lost inadvertently.
{{< /note >}}

### Backoff limit per index {#backoff-limit-per-index}

{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

{{< note >}}
You can only configure the backoff limit per index for a Job, if you
have the `JobBackoffLimitPerIndex` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
enabled in your cluster. A per-index backoff limit is only available for Indexed Jobs that
use the `restartPolicy: Never`.

> {{< /note >}}

Author: Rephrased. However, I prefer not to mix the preconditions about the cluster with the requirement for `restartPolicy: Never`. Because the `restartPolicy: Never` is not a precondition, but something that the user decides to use or not. In particular, some Indexed Jobs may use it, while other don't. 

Thus, I delete this remark entirely from the note, as the note was mixing two different things. Still, the requirement is mentioned as a comment in the sample code, next to the field, so this should be enough. It is also mentioned in the API docs for the `BackoffLimitPerIndex` field, so I think this is enough for users interested in using this field. wdyt?

> {{< /note >}}

Reviewer 3: I like this approach. Thanks!

> {{< /note >}}

Author: Great, resolving.

When you run an [indexed](#completion-mode) Job, you can choose to handle retries
for pod failures independently for each index. To do so, set the
`.spec.backoffLimitPerIndex` to specify the the maximal number of pod failures
per index.

When the per-index backoff limit is exceeded for an index, Kuberentes considers the index as failed and adds it to the
`.status.failedIndexes` field. The succeeded indexes, those with a successfully
executed pods, are recorded in the `.status.completedIndexes` field, regardless of whether you set
the `backoffLimitPerIndex` field.

> Note that a failing index does not interrupt execution of other indexes.
> You can additionally limit the maximal number of indexes marked failed by
> setting the `.spec.maxFailedIndexes` field.
> When the number of failed indexes exceeds the `maxFailedIndexes` field, the
> entire Job is marked failed by the Job controller, by setting the Failed
> condition in the Job status. This triggers termination of all running Pods for
> that Job.
> 
> Once all indexes finish for a Job where you specified a backoff limit per index,
> if at least one of those indexes did fail, the Job controller marks the overall
> Job as failed, by setting the Failed condition in the status. The Job gets
> marked as failed even if some, potentially nearly all, of the indexes were
> processed successfully.

Reviewer 3: I really like this paragraph. My only _suggestion_ would be to number the flow you are presenting:

```suggestion
Note that a failing index does not interrupt execution of other indexes.
You can additionally limit the maximal number of indexes marked failed by
setting the `.spec.maxFailedIndexes` field.

When the number of failed indexes exceeds the `maxFailedIndexes` field, the following happens:
1. The entire Job is marked failed by the Job controller, by setting the Failed
condition in the Job status. 
1. The termination of all running Pods for that Job is triggered.
1. Once all indexes finish for a Job where you specified a backoff limit per index,
if at least one of those indexes failed, the Job controller marks the overall
Job as failed, by setting the Failed condition in the status. 
1. The Job gets marked as failed even if some, potentially nearly all, of the indexes were
processed successfully.
```

Author: Rephrased to make it more clear flow, with a natural split into two scenarios, one per paragraph. The base line is when just `backoffLimitPerIndex` is specified. This is now the first paragraph. The second paragraph describes what happens when `maxFailedIndexes` is used and exceeded. wdyt?

Reviewer 3: LGTM!

Author: Great, resolving.

Here is an example manifest for a Job that defines a `backoffLimitPerIndex`:

{{< codenew file="/controllers/job-backoff-limit-per-index-example.yaml" >}}

> In the example above, the Job controller will allow for one restart for each

Reviewer 3: ```suggestion
In the example above, the Job controller allows for one restart for each
```

     should not be incremented and a replacement Pod should be created.
  - `Count`: use to indicate that the Pod should be handled in the default way.
     The counter towards the `.spec.backoffLimit` should be incremented.
  - `FailIndex`: use this action along with [backoff limit per index](#backoff-limit-per-index)

>      to avoid unnecessary retries within the index with a failed pod.

Reviewer 2: ```suggestion
     to avoid unnecessary retries within the index of a failed pod.
```
?

Author: Done

  the UNIX tool `cron`.
* Practice how to configure handling of retriable and non-retriable pod failures
  using `podFailurePolicy`, based on the step-by-step [examples](/docs/tasks/job/pod-failure-policy/).

> * Practice how to configure your Jobs to handle pod failures independently per

Reviewer 3: ```suggestion
* Learn how to configure your Jobs to handle pod failures independently per
```

Author: I actually delete this point, as this task is no longer added. We decided to leave the content for the task for later, when combined with other features around failure handling.
